common/db.py

```diff
@@ -177,24 +171,47 @@ def save_batches_to_file(
 
     def prune_old_batches(self, app_name: str, remove_border: int) -> None:
         """Helper function to prune old batches from memory."""
-        batches = self.apps[app_name]["batches"]
-        for key, batch in list(batches.items()):
-            if batch["state"] == "finalized" and batch["index"] <= remove_border:
-                del batches[key]
+        self.apps[app_name]["batches"] = {
+            batch["hash"]: batch
+            for batch in self.apps[app_name]["batches"].values()
+            if batch["state"] != "finalized" or batch["index"] > remove_border
+        }
 
 
     def get_batches(
         self, app_name: str, states: set[str], after: float = -1
     ) -> dict[str, Any]:
         """Get batches filtered by state and optionally by index."""
         batches: dict[str, Any] = {}
-        i = 0
-        for batch_hash, batch in list(self.apps[app_name]["batches"].items()):
+        loaded_batches: dict[str, Any] = {}
+        count = 0
+        last_finalized_index = self.apps[app_name]["last_finalized_batch"].get("index", 0)
+        if last_finalized_index - after >= zconfig.SNAPSHOT_CHUNK:
+            loaded_batches = self.load_finalized_batches(app_name, after + 1)
+        for batch_hash, batch in loaded_batches.items():
             if batch["state"] in states and batch.get("index", 0) > after:
                 batches[batch_hash] = batch
-                i += 1
-            if i >= zconfig.API_BATCHES_LIMIT:
+                count += 1
+            if count >= zconfig.API_BATCHES_LIMIT:
                 break
+        if count < zconfig.API_BATCHES_LIMIT and \
+           (after + 1 + count)//zconfig.SNAPSHOT_CHUNK != after//zconfig.SNAPSHOT_CHUNK and \
+           (after + 1 + count)//zconfig.SNAPSHOT_CHUNK != last_finalized_index//zconfig.SNAPSHOT_CHUNK:
+            loaded_batches = self.load_finalized_batches(app_name, after + 1 + count)
+            for batch_hash, batch in loaded_batches.items():
+                if batch["state"] in states and batch.get("index", 0) > after:
+                    batches[batch_hash] = batch
+                    count += 1
+                if count >= zconfig.API_BATCHES_LIMIT:
+                    break
+            
+        if count < zconfig.API_BATCHES_LIMIT:
+            for batch_hash, batch in list(self.apps[app_name]["batches"].items()):
+                if batch["state"] in states and batch.get("index", 0) > after:
+                    batches[batch_hash] = batch
+                    count += 1
+                if count >= zconfig.API_BATCHES_LIMIT:
+                    break
```

Comment:
This section can be implemented more cleanly.

Reply:
Done in this [commit](https://github.com/zellular-xyz/zsequencer/commit/c9d023be5410082d0bb21f89f3c157e69079fb81).